typing/typedecl.ml

```diff
@@ -2192,10 +2192,29 @@ let report_error_doc ppf = function
   | Unavailable_type_constructor p ->
       fprintf ppf "The definition of type %a@ is unavailable"
         (Style.as_inline_code Printtyp.path) p
-  | Variance Typedecl_variance.Varying_anonymous ->
-      fprintf ppf "@[%s@ %s@ %s@]"
-        "In this GADT definition," "the variance of some parameter"
-        "cannot be checked"
+  | Variance (Typedecl_variance.Varying_anonymous (n, reason)) ->
+      let reason_text =
+        match reason with
+        | Variable_constrained ty ->
+            dprintf
+              "the type variable %a appears@ in other parameters.@ \
+               Co- or contra-variant@ type parameters@ \
+               must not depend@ on other parameters."
+              Printtyp.type_expr ty
+        | Variable_instantiated ty ->
+            dprintf
+              "it is instantiated to the type %a.@ \
+               Co- or contra-variant@ type parameters@ \
+               may only appear@ as type variables@ \
+               in GADT constructor definitions."
+              Printtyp.type_expr ty
+      in
+      fprintf ppf
+        "@[<hov>In this GADT constructor definition,@ \
+         the variance of the@ %d%s parameter@ \
+         cannot be checked,@ because %t@]"
```

Comment:
Small nitpicking, I would propose to cut the `%t` at the `because` part to make the reason a full proposition:

```ocaml
 "@[<hov>In this GADT constructor definition,@ \
         the variance of the@ %d%s parameter@ \
         cannot be checked%t"
```
which makes it easier to update the reason part
```ocaml
              ",@ because the type variable %a appears@ in other parameters.@ \
               Co- or contra-variant@ type parameters@ \
               must not depend@ on other parameters."
 ```
 independently.